web3/utils/abi.py

```diff
@@ -378,8 +397,151 @@ def abi_to_signature(abi):
+        [data_tree_vals],
+    )
+
+    return pipe(data, *pipeline)
+
+
+@curry
+def abi_data_tree(types, data):
+    '''
+    Decorate the data tree with pairs of (type, data). The pair tuple is actually an
+    ABITypedData, but can be accessed as a tuple.
+
+    As an example:
+
+    >>> abi_data_tree(types=["bool[2]", "uint"], data=[[True, False], 0])
+    [("bool[2]", [("bool", True), ("bool", False)]), ("uint256", 0)]
+    '''
+    return [
+        abi_sub_tree(data_type, data_value)
+        for data_type, data_value
+        in zip(types, data)
+    ]
+
+
+@curry
+def data_tree_map(func, data_tree):
+    '''
+    Map func to every ABITypedData element in the tree. func will
+    receive two args: abi_type, and data
+    '''
+    def map_to_typed_data(elements):
+        if isinstance(elements, ABITypedData):
+            return ABITypedData(func(*elements))
+        else:
+            return elements
+    return recursive_map(map_to_typed_data, data_tree)
+
+
+@coerce_return_to_text
+def data_tree_vals(typed_data_tree):
+    '''
+    Strip all types out of the data tree
+    '''
+    data_tree = recursive_map(strip_abi_type, typed_data_tree)
+    if len(data_tree) == 1:
+        return data_tree[0]
+    else:
+        return data_tree
+
+
+class ABITypedData(namedtuple('ABITypedData', 'abi_type, data')):
+    '''
+    This class marks data as having a certain ABI-type.
+
+    >>> a1 = ABITypedData(['address', addr1])
+    >>> a2 = ABITypedData(['address', addr2])
+    >>> addrs = ABITypedData(['address[]', [a1, a2])
+
+    You can access the fields using tuple() interface, or with
+    attributes:
+
+    >>> assert a1.abi_type == a1[0]
+    >>> assert a1.data == a1[1]
+
+    Unlike a typical `namedtuple`, you initialize with a single
+    positional argument that is iterable, to match the init
+    interface of all other relevant collections.
+    '''
+    def __new__(cls, iterable):
+        return super(ABITypedData, cls).__new__(cls, *iterable)
+
+
+def abi_sub_tree(data_type, data_value):
+    try:
+        base, sub, arrlist = data_type
+    except ValueError:
+        base, sub, arrlist = process_type(data_type)
+
+    collapsed = collapse_type(base, sub, arrlist)
+
+    if arrlist:
+        sub_type = (base, sub, arrlist[:-1])
+        return ABITypedData([
+            collapsed,
+            [
+                abi_sub_tree(sub_type, sub_value)
+                for sub_value in data_value
+            ],
+        ])
+    else:
+        return ABITypedData([collapsed, data_value])
+
+
+# TODO if this makes any sense at all, it probably belongs in eth_abi
+def collapse_type(base, sub, arrlist):
+    return str(base + sub + ''.join(map(repr, arrlist)))
+
+
+def strip_abi_type(elements):
+    if isinstance(elements, ABITypedData):
+        return elements.data
+    else:
+        return elements
+
+
 @coerce_return_to_text
 def normalize_return_type(data_type, data_value):
+    # TODO replace all uses with map_abi_data
```

Comment:
Is this still a `TODO`?

Reply:
Yeah, I was going to punt it, but I'll go add it now.

Reply:
Glad I did. I realized the obvious, when applying the same pattern to events (that popping single elements out of their list is a special case). I removed the special case from `map_abi_data` and into the contract code.